calico/src/main/scala/calico/html/Modifier.scala

```diff
@@ -36,12 +36,8 @@ trait Modifier[F[_], E, A]:
 
 private object Modifier:
   def forSignal[F[_]: Async, E, M, V](signal: M => Signal[F, V])(
-      mkModify: (M, E) => V => F[Unit]): Modifier[F, E, M] = (m, e) =>
-    signal(m).getAndUpdates.flatMap { (head, tail) =>
-      val modify = mkModify(m, e)
-      Resource.eval(modify(head)) *>
-        tail.foreach(modify(_)).compile.drain.cedeBackground.void
-    }
+      mkModify: (M, E) => V => F[Unit]): Modifier[F, E, M] =
+    forSignalResource(signal.andThen(_.pure))(mkModify)
```

Comment:
Plus a new Function1 I guess :smile: 

In general I avoid Premature Optimization, but I don't have a feeling yet for how far away we are from "too slow". So, if you are concerned, I'll close it.

Reply:
I'm not sure either, but we are definitely slowest at the moment 😅 

If you want a sobering moment, check out this performance comparison.
https://armanbilge.github.io/react-angular-ember-elm-performance-comparison

Reply:
:flushed: seems like we have a pretty tiny bottleneck to find. Any ideas already? Stream?

Reply:
![image](https://user-images.githubusercontent.com/1238350/214378491-c893f4f6-bff8-4136-9f68-4ac3eeda64af.png)

Reply:
I'm not sure if it's anything specific. Programs-as-values, Cats Effect, FS2, they all have some overheads. We are drowning in allocations.

Also remember FS2 wraps everything in `Chunk`, which is kind of pointless for us 😅 and also because FS2 is pull-based, that means there is a two-way interaction every time an event fires (upstream sends event to downstream, downstream resubscribes to upstream).

I've been doing some work in Cats Effect lately to try and improve the situation. E.g.
- https://github.com/typelevel/cats-effect/pull/3222
- https://github.com/typelevel/cats-effect/pull/3228
- https://github.com/typelevel/cats-effect/pull/3281
- https://github.com/typelevel/cats-effect/pull/3324

And also trying to avoid allocations as much as possible.
- https://github.com/armanbilge/fs2-dom/pull/33

But at some level it's fundamental.

Reply:
Did these changes improve performance and if so, how much?

Reply:
I don't really know. The only one I properly measured was https://github.com/typelevel/cats-effect/pull/3225, which saw a 60% performance improvement. But that was measuring a Node.js server, so it's not easy to translate that result :)

The problem is we don't really have good benchmarking infrastructure for JS. I started working on a project to fill that gap but haven't made good progress on it recently.
- https://github.com/armanbilge/scala-jmh

Reply:
> The only one I properly measured was https://github.com/typelevel/cats-effect/pull/3225, which saw a 60% performance improvement.

Impressive! I wonder about the potential of batching microtasks, since my quick look into dev tools suggests that there could be something to get. But I guess it's probably hard to find the right balance in regards to rendering time.

Reply:
> I wonder about the potential of batching microtasks

When we upgrade to CE 3.5.0, Calico will no longer use microtasks.

We are currently using microtasks as a workaround for the shortcomings of macrotasks.

https://armanbilge.github.io/calico/#task-scheduling-and-glitch-free-rendering

Fortunately the new `BatchingMacrotaskExecutor` solves all those problems in a much better way.

Reply:
> When we upgrade to CE 3.5.0, Calico will no longer use microtasks

That'll be exciting to "bench". Do you have a current repo fork of https://github.com/evancz/react-angular-ember-elm-performance-comparison with calico included?

Reply:
https://github.com/armanbilge/react-angular-ember-elm-performance-comparison/tree/gh-pages

It benches the live TodoMVC.
https://armanbilge.github.io/calico/todomvc/